File: Tests/RangerTests.cs
using System;
using Stringier.Patterns;
using static Stringier.Patterns.Pattern;
using Xunit;

namespace Tests {
	public class RangerTests {
		[Fact]
		public void Simple_Consume() {
			Pattern pattern = Range("Hello", ';');
			ResultAssert.Captures("Hello;", pattern.Consume("Hello;"));
			ResultAssert.Captures("Hello World;", pattern.Consume("Hello World;"));
		}

		[Fact]
		public void Simple_Neglect() => Assert.Throws<PatternConstructionException>(() => Not(Range("Hello", ";")));

		[Fact]
		public void Escaped_Consume() {
			Pattern pattern = Range("\"", "\"", "\\\"");
			ResultAssert.Captures("\"\"", pattern.Consume("\"\""));
			ResultAssert.Captures("\"H\"", pattern.Consume("\"H\"i"));
			ResultAssert.Captures("\"Hello\"", pattern.Consume("\"Hello\""));
			ResultAssert.Captures("\"Hello\\\"Goodbye\"", pattern.Consume("\"Hello\\\"Goodbye\""));
		}

		[Fact]
		public void Nested_Consume() {
			Pattern pattern = NestedRange("if", "end if");
			ResultAssert.Captures("if\nif\nend if\nbacon\nend if", pattern.Consume("if\nif\nend if\nbacon\nend if\nfoobar"));
			ResultAssert.Fails(pattern.Consume("if\nif\nend if\nbacon\nfoobar"));
		}
	}
}

File: Tests.FSharp/RangerTests.fs
namespace Patterns

open System
open Stringier.Patterns
open Defender
open Xunit

type RangerTests() =
    inherit Trial()

    [<Fact>]
    member _.``range consume`` () =
        range "Hello" ';'
        |> consumes "Hello;" "Hello;"
        |> consumes "Hello World;" "Hello World;"
        |> ignore

    [<Fact>]
    member _.``erange consume``() =
        erange "\"" "\"" "\\\""
        |> consumes "\"\"" "\"\""
        |> consumes "\"H\"" "\"H\""
        |> consumes "\"Hello\"" "\"Hello\""
        |> consumes "\"Hello\\\"Goodbye\"" "\"Hello\\\"Goodbye\""
        |> ignore

    [<Fact>]
    member _.``nrange consume`` () =
        nrange "if" "end if"
        |> consumes "if\nif\nend if\nbacon\nend if" "if\nif\nend if\nbacon\nend if\nfoobar"
        |> fails "if\nif\nend if\nbacon\nfoobar"
        |> ignore